EssentialEstablishmentGenerator/Town/js/createTown.js

```diff
@@ -216,15 +217,43 @@ setup.createTown = function (base) {
   Object.keys(town.roll).forEach(function (roll) {
     town.roll[roll].clamp(1, 100)
   })
-
+  town.townMaterial = createTownMaterial(setup.townData.terrain[town.terrain].location[town.location].possibleMaterials, town.roll.wealth, town.roll.size)
   setup.townRender(town)
   setup.createStartBuildings(town)
   setup.createStartFactions(town)
-
+  setMaterialProbability(setup.townData.terrain[town.terrain].location[town.location].possibleMaterials)
   console.log(town)
   console.groupEnd()
   // setup.createWeather(town)
   console.log(town.name + ' has loaded.')
-
+  console.log('loaded', town.wealth, town.roll.size)
+  confirm(town.townMaterial)
   return town
 }
+
+function createTownMaterial (startLocationMaterials, wealth, size) {
+  const filterArrayMaterial = (material) => {
+    return startLocationMaterials.includes(material)
+  }
+  const highTierMaterials = ['plaster', 'bricks', 'limestone', 'gypsum'].filter(filterArrayMaterial)
+  const midTierMaterials = ['hewn rock', 'stone', 'cobblestone', 'wood'].filter(filterArrayMaterial)
+  const lowTierMaterials = ['wood', 'adobe', 'daub', 'cob', 'straw', 'rock', 'terra cotta', 'clay', 'cobblestone'].filter(filterArrayMaterial)
+  const materialDecider = wealth + (size * 0.1)
+  if (materialDecider >= 70) {
+    return highTierMaterials.seededrandom()
+  } else if (materialDecider >= 50 && materialDecider < 70) {
+    return midTierMaterials.seededrandom()
+  } else if (materialDecider <= 49) {
+    return lowTierMaterials.seededrandom()
+  }
+}
+
+function setMaterialProbability (startLocationMaterials) {
+  Object.keys(setup.structure.material).forEach((material) => {
+    startLocationMaterials.forEach(element => {
+      if (element !== material) {
+        setup.structure.material[material].probability = 0
```

Comment:
Typically, it is bad practice to mutate `setup` objects. They are meant to be entirely static, and in the Twine build, they do not take kindly to being mutated (although this will not be an issue for too long, as we'll be shifting over to React). This is probably one of the reasons that `probability = 0` is still returning them for you.

Reply:
I see, so shifting it to `setup`, like you suggested above would solve this issue ?

Reply:
No. Function evaluations should never mutate `setup`; setup is purely for information storage. if you feel that that is the best way to implement it you could make a clone of `setup.structure.material.types` that would be trashed automatically after the function had `return`ed. You could also refactor to evaluate upon runtime, passing through the `town` object. Remember that the intent of the Generator is that it can be modified; the user may change the biome, making the dominant building material different.